JP Onboarding: Increase delay and attempts in retryPolicy of Save action

ockham opened the pull request:
If WooCommerce installation times out on the self-hosted site (because it was too slow to fetch the plugin via `curl` from WordPress.org servers), instead of showing an error message, retry two more times. The `POST` action we're performing won't break anything upon subsequent attempts (and is actually idempotent AFAICT), and plugin fetching and installation might have been successful by the time we try again, and thus return a 'success' code.

I'm essentially mimicking behavior that the data-layer's [`retry-on-failure` ](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer/wpcom-http/pipeline/retry-on-failure) would give us, if we [weren't dealing](https://github.com/Automattic/wp-calypso/blob/f8b29158dc801399f4b91a32cda9cef1fc94bffd/client/state/data-layer/wpcom-http/pipeline/retry-on-failure/index.jsx#L23-L26) with a `POST` request here.

Fixes #21929.

To test:
* You need a self-hosted WP install that somewhat reproducibly times out on WooCommerce install during JPO: Pressable seems to do the trick.
* Verify on `master` or wpcalypso/staging that you get the error documented in #21929.
* Check out this branch in your local Calypso dev environment.
* Make sure WooCommerce is not installed on your self-hosted site.
* Start the JPO flow at https://yourslow.mystagingwebsite.com/wp-admin/admin.php?page=jetpack&action=onboard&calypso_env=development
* When prompted, select 'Business Site' as site type.
* Skip thru to the WooCommerce installation step, click the 'Yes' button there.
* On the summary page, wait a little. You shouldn't see any error notice. Then, check Redux dev tools: among the last actions dispatched, there should be a`JETPACK_ONBOARDING_SETTINGS_SAVE_SUCCESS` action with `settings: {installWooCommerce: true }` payload, indicating that installation was successful.
* On your site's `wp-admin`, verify that WooCommerce has been successfully installed and activated.

## Possible alternative approach:
This PR doesn't do a very faithful job in mimicking `retry-on-failure` (e.g. when it comes to timeouts), and it might not make sense to actually copy that fancy implementation over b/c redundancy sucks. Maybe it's instead worth considering adding a flag to `retry-on-failure` to enable retries for `POST` requests. I think the relevant criterion is idempotency, so I'd probably make it a boolean flag called `idempotent` in `meta.dataLayer`.

ockham commented:
cc @dmsnell in case I'm missing something obvious...

dmsnell commented:
The `GET` check [occurs](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/pipeline/retry-on-failure/index.jsx#L24) outside of the policies. This was done to protect against replaying non-safe POST requests.

If we want to override that behavior then we'd need to change the behavior of the pipeline stage and move the request method into the policies somehow.

As discussed in Slack, the alternative is handling the POST failures in the data layer handlers, which has the advantages of being explicit on retry and of disambiguating the types of failures.

So far we haven't had ned to turn the policies into functions that behave differently based on the request because it's been pretty constrained to only know about HTTP and not about Redux, but I could imagine sending a few parameters in there like HTTP method and existing retry count which could result in returning different delays and whether or not to retry at all.

ockham commented:
Thanks Dennis! Okay, going to implement it in `onFailure` for now.

ockham commented:
Ready for review!

Would like to hear your opinion @dmsnell, especially on the 'Possible alternative approach' I'm discussing in the PR desc.

ockham commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+
+	// We cannot use `extendAction( action, ... )` here, since `meta.datayLayer` now includes error information,
+	// which we would propagate, causing the data layer to think there's been an error on the subsequent attempt.
+	// Instead, we have to re-assemble our action.
```
Should there be a generic tool that strips error information from `meta.dataLayer` for cases like this?

dmsnell commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+	// since it might just be a slow server that actually ends up installing it
+	// properly, in which case a subsequent request will return 'success'.
+	if (
+		settings.installWooCommerce !== true ||
```
could `settings` be `null` or `undefined` here?

tyxla replied:
I agree with Dennis here; this could potentially happen in cases where we're not installing WooCommerce, so we might want to tighten this one.

dmsnell commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+export const retryOrAnnounceSaveFailure = ( { dispatch }, action ) => {
+	const MAX_WOOCOMMERCE_INSTALL_RETRIES = 2;
+	const { settings, siteId, type, meta: { dataLayer } } = action;
+	const { error, retryCount = 0 } = dataLayer;
```
`error` is already coming in as the third argument to this function

tyxla commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+};
+
+export const retryOrAnnounceSaveFailure = ( { dispatch }, action ) => {
+	const MAX_WOOCOMMERCE_INSTALL_RETRIES = 2;
```
We don't need to define this one on every function call; rather we can move it outside, and possibly export it to use it in tests.

tyxla commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+	// properly, in which case a subsequent request will return 'success'.
+	if (
+		settings.installWooCommerce !== true ||
+		error.error !== 'http_request_failed' ||
```
I don't think this is covering our use case well enough. This will inadvertently handle a bunch of other failure cases that fail with a `http_request_failed` error, for example (and that's not a full list):

* Could not open handle for fopen() to /path/to/file
* Too many redirects.
* Failed to write request to temporary file.
* A valid URL was not provided.
* User has blocked requests through HTTP.
* Destination directory for file streaming does not exist or is not writable.
* Any generic cURL error that came from `curl_error()`

I'd rather check for presence of `cURL error 28` in the error message, so we can only do this for cases when it actually times out.

tyxla commented on client/state/data-layer/wpcom/jetpack-onboarding/index.js:
```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
+		return announceSaveFailure( { dispatch }, { siteId } );
+	}
+
+	// We cannot use `extendAction( action, ... )` here, since `meta.datayLayer` now includes error information,
```
Typo here, `datayLayer` should be `dataLayer`

ockham replied:
Ay!

ockham commented:
Addressed feedback. @tyxla Wanna give this a final 👍?

> We missed this apparently when it went in, but the handler for `/jetpack/v4/settings` should probably live in `state/data-layer/wpcom-http/jetpack/settings/index.js`
>
> are you aware of any reasons why it was put here instead?

@dmsnell Oversight on our part, I guess -- we were naming folders after corresponding ones in `state/`. However, I'm fine with renaming this (in a subsequent PR).

@dmsnell and @tyxla I'd also be curious to hear both your opinions on the 'Possible alternative approach' in the PR desc, and on https://github.com/Automattic/wp-calypso/pull/22454#discussion_r168818656 (non-blocking to this PR however).

dmsnell commented:
@ockham I think this is probably a good compromise for now. concerning the `PSOT` flag I don't feel good about that option. Instead I'd rather move the `GET` check into the policies and compose their behavior, changing their signatures a little bit too so they could examine the action.

that would make the default policies compositions of something like `exponentialBackoff . onlyGET` while this one would become `exponentialBackoff` or `exponentialBackoff . explicitPOST`

if you are willing to wait for it I think the result would be better

dmsnell commented:
No need to revert! This patch is good and sufficient on its own (as long as it's not causing bugs from other logic, like trapping events it shouldn't)

I only meant to comment on the alternative solution you proposed.
